include/options.hxx

```diff
@@ -388,7 +420,7 @@ public:
                   << ")" << std::endl;
       return def;
     }
-    T val = as<T>();
+    T val = as<T>(bout::utils::meshFromValue(def));
```

Comment:
It looks like the signature for `as` is still `as<T>(Mesh* UNUSED(mesh) = nullptr)` -- should it now be using `mesh`?

Reply:
The `<Field2D>` and `<Field3D>` specialisations do use `mesh`, but the generic version does not. Really it would be nice to just have an optional `T`, meaning "make one like this": This would make it easier to handle things like staggering location or other field attributes that we might want to pass through.

Is there a nice way to do this? Perhaps something like
```
as<T>(const T& similar_to = {})
```
Can a default initialiser work for arbitrary type?

Reply:
Oops, didn't spot that! Thanks!

Yes, I think you want:

```cpp
as<T>(T&& similar_to = {})
```

This should avoid copies when passing values but also allow rvalues: https://godbolt.org/z/zPczXB

Reply:
We've also now got `emptyFrom` for fields that might be useful here

Reply:
Great thanks! Any objection to changing this API to the above `T&& similar_to = {}`? I don't think this functionality is used anywhere apart from tests and examples (yet).

Reply:
Nope, fine by me!

Reply:
Great! I think with that change I could remove the `meshFromValue` hack, and the `emptyFrom` function might come in handy. Thanks!

Reply:
Quick semi-related question: In a couple of places I now have something like:
```
Field3D result {emptyFrom(similar_to)}; // Make a field with correct metadata
result = scalar_value; // Fill with the scalar value
return result;
```
would it be nicer to have something like
```
return emptyFrom(similar_to).fill(scalar_value);
```
or 
```
return fill(emptyFrom(similar_to), scalar_value);
```
?

Reply:
We made a `zeroFrom` function as well as `emptyFrom` but decided at the time not to add anything for non-zero values because we didn't have a place where that was needed. Looks like that's changed now! I quite like the idea of having a `.fill(scalar_value)` method, it goes nicely with the method-chaining approach to changing properties (e.g. `emptyFrom(f).setLocation(location)`), and if we add it, we could get rid of `zeroFrom` :+1:

Reply:
I am conflicted -- the method (first) is definitely easier to read, but the free function is cleaner. `fill` doesn't need to know anything about the field's internals, and is simple enough to implement as a template that works for basically anything.

As a third option, what about something like
```cpp
return filledFrom(similar_to, scalar_value);
```
?

For this particular case, you could even do the (less readable):
```cpp
return emptyFrom(similar_to) = scalar_value;
```
I'm not sure I like that though! :smile:

Reply:
Perhaps `filledFrom` would be good. I am also conflicted about the method vs free function, but this might be the cleanest way.